sources/Debugger/TileViewer/TileViewer.cpp

```
@@ -0,0 +1,149 @@
+	}
+
+	void TileViewer::disableViewer()
+	{
+		this->_snes.disableTileViewerDebugging();
+	}
+
+	void TileViewer::focus()
+	{
+		this->_window->activateWindow();
+	}
+
+	bool TileViewer::isDebugger()
+	{
+		return true;
+	}
+
+	uint16_t TileViewer::read(uint8_t addr)
+	{
+		return this->_ppu.cgramRead(addr);
+	}
+
+	void TileViewer::setPaletteIndex(int paletteIndex)
+	{
+		this->_ramTileRenderer.setPaletteIndex(paletteIndex);
+		this->internalUpdate();
+	}
+
+	void TileViewer::setBpp(int bpp)
+	{
+		this->_ui.PaletteIndex->setDisabled(bpp > 4);
+		switch (bpp) {
+		case 8:
+			this->_ui.PaletteIndex->setValue(0);
+			break;
+		case 4:
+			this->_ui.PaletteIndex->setMaximum(15);
+			if (this->_ui.PaletteIndex->value() > 15) {
+				this->_ui.PaletteIndex->setValue(15);
+			}
+			break;
+		case 2:
+		default:
+			bpp = 2;
+			this->_ui.PaletteIndex->setMaximum(63);
+			break;
+		}
+		this->_ramTileRenderer.setBpp(bpp);
+		this->internalUpdate();
+	}
+
+	void TileViewer::setNbColumns(int nbColumns)
+	{
+		this->_ramTileRenderer.setNbColumns(nbColumns);
+		this->internalUpdate();
+	}
+
+	void TileViewer::setRenderSize(int size)
+	{
+		this->_ramTileRenderer.setRenderSize(size);
+		this->internalUpdate();
+	}
+
+	int TileViewer::getBpp() const
+	{
+		return this->_ramTileRenderer.getBpp();
+	}
+
+	int TileViewer::getPaletteIndex() const
+	{
+		return this->_ramTileRenderer.getPaletteIndex();
+	}
+
+	int TileViewer::getNbColumns() const
+	{
+		return this->_ramTileRenderer.getNbColumns();
+	}
+
+	void TileViewer::internalUpdate()
+	{
+		this->_ramTileRenderer.render();
+		this->_sfWidget->buffer = this->_ramTileRenderer.buffer;
+	}
+
+	void TileViewer::setRamOffset(int offset)
+	{
+		this->_ramTileRenderer.setRamOffset(offset);
+		this->internalUpdate();
+	}
+
+	void TileViewer::_bppChangeUIHandler(int index)
+	{
+		switch (index) {
+		case 0: return this->setBpp(2);
+		case 1: return this->setBpp(4);
+		case 2: return this->setBpp(8);
+		default:
+			break;
```

Please throw on unexpected threads.

---

fixed

---

fixed